internal/k8s/reconciler/statuses.go

```diff
@@ -0,0 +1,84 @@
+package reconciler
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/hashicorp/go-hclog"
+
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/gatewayclient"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/reconciler/status"
+	"github.com/hashicorp/consul-api-gateway/internal/store"
+)
+
+var _ store.StatusUpdater = (*StatusUpdater)(nil)
+
+type StatusUpdater struct {
+	logger         hclog.Logger
+	client         gatewayclient.Client
+	deployer       *GatewayDeployer
+	controllerName string
+}
+
+func NewStatusUpdater(logger hclog.Logger, client gatewayclient.Client, deployer *GatewayDeployer, controllerName string) *StatusUpdater {
+	return &StatusUpdater{
+		logger:         logger,
+		client:         client,
+		deployer:       deployer,
+		controllerName: controllerName,
+	}
+}
+
+func (s *StatusUpdater) UpdateGatewayStatusOnSync(ctx context.Context, gateway store.Gateway, sync func() (bool, error)) error {
```

Comment:
This code originates in `K8sGateway.TrackSync( ... )`. The original code will be removed in a followup PR.